crates/ra_hir_expand/src/db.rs

```diff
@@ -97,8 +132,40 @@ pub(crate) fn parse_macro(
             log::warn!("fail on macro_parse: (reason: {})", err,);
         })
         .ok()?;
-    match macro_file.macro_file_kind {
-        MacroFileKind::Items => mbe::token_tree_to_items(&tt).ok().map(Parse::to_syntax),
-        MacroFileKind::Expr => mbe::token_tree_to_expr(&tt).ok().map(Parse::to_syntax),
-    }
+
+    let (parsed, exp_map) = match macro_file.macro_file_kind {
+        MacroFileKind::Items => {
+            mbe::token_tree_to_items(&tt.0).map(|(p, map)| (p.to_syntax(), map)).ok()?
+        }
+        MacroFileKind::Expr => {
+            mbe::token_tree_to_expr(&tt.0).map(|(p, map)| (p.to_syntax(), map)).ok()?
+        }
+    };
+
+    let expand_info = tt.1;
+    let loc: MacroCallLoc = db.lookup_intern_macro(macro_call_id);
+
+    let arg_tt = loc.ast_id.to_node(db).token_tree();
+    let def_tt = loc.def.ast_id.to_node(db).token_tree();
+
+    let arg_range = arg_tt.map(|t| t.syntax().text_range());
+    let def_range = def_tt.map(|t| t.syntax().text_range());
+
+    let shift = db.macro_def(loc.def)?.0.shift();
+
+    let arg_map =
+        arg_range.map(|it| exp_map.map_ranges(&expand_info.arg_map, it, shift)).unwrap_or_default();
+    let def_map =
+        def_range.map(|it| exp_map.map_ranges(&expand_info.def_map, it, 0)).unwrap_or_default();
```

Comment:
I feel like most of `.map`s in this function can still be safetly replaced with `?`. If we can't even get a tt out of macro, we surely shoudn't be able to parse it

Reply:
Um.. do I missing something ? some of the given input tokens will not be able to find in output tokens, e.g:
```rust
macro_rules foo!($a:ident) => { "hello"}
foo!(b) // output "hello"
```
Here token `b` would return to `None` in `map_ranges`, right ?

Reply:
I thnk

```rust
let arg_tt = loc.ast_id.to_node(db).token_tree();
```

can be replaced with

```rust
let arg_tt = loc.ast_id.to_node(db).token_tree()?;
```

Though, I suspect if we avoid constcructing an `ExpansionMap` here and just return a `RefDefMap`, this code might go away naturally?

Reply:
Yes, kind of , right now the logic implement here is moved to `HirFileId::parent_expansion`.